### ISSUE TITLE
Ansible: Consider switching to Python3 install

### ISSUE THREAD

**Issue by JedMeister** [turnkeylinux/tracker#1300]
Whilst Python2 will still be included in the next Debian release, it is currently on life support really...

Most people have switched to using Python 3 these days and almost no new development is being done in Python 2. As also noted in #887 nobody who doesn't already know/use Python 2 is learning it (new users are starting with Python 3).

As such, it seems to me that installing Ansible for Python 3 makes plenty of sense. Having said that, perhaps it is possible to provide both and allow the user to choose which they use (or perhaps even support both?)

@Dude4Linux - Seeing as you're an Ansible user (not to mention the main developer of this appliance) I'd love your thoughts and input.

**Comment by Dude4Linux** [comment]
@JedMeister - I agree that we need to focus on conversion of all appliances to Python3 in the next release. This came up as I was working on a new appliance for Mautic (social media marketing automation).  Not being familiar with either Python 2.7 or 3.0, it wasn't clear to me how I could accomplish this with TKLDev.  I think we will need some tools to audit our existing appliance code and identify programs that need updating.  I don't know of any such tools, but I bet they exist.

**Comment by Dude4Linux** [comment]
@JedMeister - Since the `default` /usr/bin/python is set using a softlink and not /etc/alternatives, maybe just redirect that link to python3.5 and see what breaks.  Most applications have or are being upgraded to python3.  If there are any apps that still require python2.7, adjustments would have to be made to specifically call out /usr/bin/python2.
```
# ll /usr/bin/python*
lrwxrwxrwx 1 root root       9 Jan 24  2017 /usr/bin/python -> python2.7*
lrwxrwxrwx 1 root root       9 Jan 24  2017 /usr/bin/python2 -> python2.7*
-rwxr-xr-x 1 root root 3779512 Sep 26 18:42 /usr/bin/python2.7*
lrwxrwxrwx 1 root root       9 Jan 20  2017 /usr/bin/python3 -> python3.5*
-rwxr-xr-x 2 root root 4751184 Sep 27 16:03 /usr/bin/python3.5*
-rwxr-xr-x 2 root root 4751184 Sep 27 16:03 /usr/bin/python3.5m*
lrwxrwxrwx 1 root root      10 Jan 20  2017 /usr/bin/python3m -> python3.5m*
```

**Comment by JedMeister** [comment]
As you note, Debian Stretch (i.e. v15.x) provides both Python 2 (2.7) and Python 3 (3.5). FWIW both being installed by default accounts for a significant part of ISO size increase between v14.x and v15.x - Python 3 was optional in v14.x and only installed when done so explicitly.

By design, when writing scripts, the version you use is controlled by the shebang (or explicitly calling python2/python3 - e.g. `python3 /path/to/py3-script.py`). Ideally `#!/usr/bin/python2` is used for python2 scripts, whereas `#!/usr/bin/python3` is required when using python3.

As per your research, in Debian (and all 'Nix OS, with exception of Arch Linux), currently `python` (aka `/usr/bin/python`) points to python2 by default (and is expected behaviour - at least for python 'Nix users). It's worth noting that as per [PEP 394](https://www.python.org/dev/peps/pep-0394/), best practice is to only use the `#!/usr/bin/python` shebang when the script is compatible with both py2 & py3. I don't think that we can trust/assume that Debian does that though.

As noted in the above PEP, Arch Linux is the only 'Nix to default `python` -> `python3` so I don't think changing that is the right approach. It will likely break many of the Debian built in OS components (many "low level" OS scripts are still only available in python2, while many newer ones are python3 - hence why both versions are installed/required by default). At best it will likely introduce weird bugs. Besides, it would likely upset any TurnKey user who knows anything about Python on 'Nix.

However, for some specific tools, such as `pip`, it may be worth considering using the 'alternatives' system to point `pip` -> `pip3` (the command for the Debian package of pip for python3). However, we should be careful to ensure that if the user explicitly installs `python-pip` that is what they get if they run `pip`.

WRT to Debian packages, they are all either explicitly python2 or 3. The `python-<py-lib-name>` is the python2 variant, whereas the `python3-<py-lib-name>` is the python3 one. As hinted above, if you wish to use the Debian py3 pip to install additional py3 libs, then you'll first want to install `python3-pip` then install additional modules via `pip3 <py-lib-name>`.

**Comment by Dude4Linux** [comment]
The guidance from Ansible re: Python3 is pretty clear.  Ansible has supported Python2&3 since version 2.5.  To use v3 on the server you must install it with pip3.  Ansible will continue to use Python2 on the target hosts unless the playbook developer specifies Python3 (host by host basis).  It seems to me that switching to pip3 on the next release and addending the documentation should be sufficient.

**Comment by JedMeister** [comment]
Sounds good @Dude4Linux.

Assuming that we don't include python-pip (i.e. python2 pip package) by default then perhaps we could create a little wrapper/helper script to make life a little easier for Linux python newbs? I suspect that they might expect `pip` to work for python3 OOTB? (And think it's not installed if it doesn't work).

What do you think? FWIW I have an idea that I could write up (at least as an example) if you think my rationale makes sense and it's worth the effort...?!

**Comment by Dude4Linux** [comment]
@JedMeister - I've been able to get Ansible installed with Python3 by adding the python3-modules needed by Ansible.  I added python3-pip to plan/main.  It's no longer necessary to do some funky stuff in `conf.d/main` to install the latest version of `pip`.  I left the Python2 version of all modules including pip installed for now.  Once I've tested everything, I'll look into removing them.  I did find that I had to install the v3 module for `wheel`.  (the v2 wheel module is inherited from either core or common)
I don't recommend redirecting `pip` to `pip3`.  I think that would confuse users and would break scripts that try to use it.  For the interim transition I'd recommend installing both `pip` and `pip3` although it will increase the image size.

From `root.patched`
```
# ansible --version
ansible 2.5.9
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/library', '/usr/share/ansible/library']
  ansible python module location = /usr/local/lib/python3.5/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.5.3 (default, Sep 27 2018, 17:25:39) [GCC 6.3.0 20170516]
```
TODO:  There are three python2 files that are included in `overlays`.  I've been experimenting converting them using `futurize` so they are compatible with either v2 or v3 following the procedure outlined in
http://python-future.org/overview.html  and http://python-future.org/automatic_conversion.html
I'll let you know how that works out.  I have found that I can't test `ansible.py` with Python3 until the underlying infrastructure (inithooks & turnkey-pylib) have been updated.  Do you have a plan and schedule for when that will happen? 
FWIW:  I'm thinking there needs to be a transition period when both python2 and python3 are supported. Then in 2020 when python2 is no longer supported, we can finally remove it.  Some applications may have to be deprecated if they haven't been upgraded to v3.

**Comment by Dude4Linux** [comment]
@JedMeister - So I've been able to build a container image for Ansible running on Python3 as described above.  Running `futurize` against `ansible.py`, the program that changes the ansible user's email and password, produced the following changes:
```
 diff ansible.py.bak ansible.py
12a13
> from __future__ import print_function
24c25
<     print >> sys.stderr, "Error:", s
---
>     print("Error:", s, file=sys.stderr)
29,31c30,32
<         print >> sys.stderr, "Error:", s
<     print >> sys.stderr, "Syntax: %s <username> [options]" % sys.argv[0]
<     print >> sys.stderr, __doc__
---
>         print("Error:", s, file=sys.stderr)
>     print("Syntax: %s <username> [options]" % sys.argv[0], file=sys.stderr)
>     print(__doc__, file=sys.stderr)
38c39
<     except getopt.GetoptError, e:
---
>     except getopt.GetoptError as e:
```
No other changes were necessary.  I can only run it with `python2` since the libraries it depends on haven't been updated.  In theory, it will be compatible with both versions and will continue to work when the underlying infrastructure is updated.

The `turnkey_facts.py` module is another story.  Running `futurize` seemed to break more things than it fixed.  I was forced to refer to Ansible's documentation on writing modules compatible with both python 2 & 3.  There was also the issue of needing to update the module to be compatible with the newer versions of Ansible.  I have been able to do limited testing to make sure it is working with Python2.

I haven't yet tackled the third program `lxc_ssh.py` because I don't have a clean way to test it in my mobile environment.  This may have to wait until I return home.

**Comment by JedMeister** [comment]
@Dude4Linux 

> I've been able to get Ansible installed with Python3 by adding the python3-modules needed by Ansible. I added python3-pip to plan/main. It's no longer necessary to do some funky stuff in conf.d/main to install the latest version of pip.

Awesome! :+1: 

> I left the Python2 version of all modules including pip installed for now. Once I've tested everything, I'll look into removing them. I did find that I had to install the v3 module for `wheel`. (the v2 wheel module is inherited from either core or common)

Ah ok. Well, so long as it's included in the plan, that shouldn't be an issue.

> I don't recommend redirecting `pip` to `pip3`. I think that would confuse users and would break scripts that try to use it. 

FWIW, what I had in mind was a wrapper script (in `/usr/local/bin`) that would automatically (and probably silently) pass the command and arguments to `/usr/bin/pip` if it existed (i.e. users with python2 pip installed and wishing to use that). Otherwise would show a notice via `stderr` (which could be disabled by the user if they wished) about the fact that it was passing the command and switches to `pip3` (i.e. `/usr/bin/pip3`).

I think an approach like that would mitigate your concerns. OTOH, it may still cause edge case issues and does add a degree of complexity (not to mention be unexpected by many users). And if we do still include (py2) pip, then it doesn't make any sense at all.

So I'm happy to scratch that idea and just document the use of pip3 and educate any users who have issues.

> For the interim transition I'd recommend installing both `pip` and `pip3` although it will increase the image size.

TBH, I'm not clear on the rationale for inclusion of py2 pip if we're only installing Ansible for Python3?

If we're installing Ansible for both py2 and py3 then it certainly makes sense (but is that what we want to do?). Iif we only provide Ansible py3, then my inclination would be to not include any additional py2 packages (beyond what we normally ship).

But I'm not firm on that and am certainly open to further discussion.

> TODO: There are three python2 files that are included in `overlays`. I've been experimenting converting them using `futurize` so they are compatible with either v2 or v3 following the procedure outlined in
http://python-future.org/overview.html and http://python-future.org/automatic_conversion.html

FWIW I've used `2to3` with great success (although I note that it's not always perfect). Although if you're trying to maintain py2 compatibility (which certainly makes sense if we'll include Ansible for both py2 and py3) then that isn't really relevant.

> I'll let you know how that works out. I have found that I can't test `ansible.py` with Python3 until the underlying infrastructure (inithooks & turnkey-pylib) have been updated.

You mean the [Ansible inithook](https://github.com/turnkeylinux-apps/ansible/blob/master/overlay/usr/lib/inithooks/bin/ansible.py) I assume? If so, then (at least on face value) there shouldn't be any need to update that? Although perhaps you are leveraging it from within one of the other scripts? Even if you do though, you could simply call it via the commandline (rather than importing it as a module).

> Do you have a plan and schedule for when that will happen?

Considering that python2 will still be available and supported in Buster/v16.0, there is no imperative to do that (until we get to Bullseye - what one would assume to be the basis of v17.x; which probably won't have py2 anymore). 

Although OTOH, out of interest, I tried removing python2 from a v15.x TKLDev instance and with one exception, the only things that require python2 (even in Stretch) are our tools! And even the exception (`etckeeper`) is a dependency bug (it isn't even written in python!). I didn't dig very deep, but it does seem that already we are behind the 8-ball...

I think that we should be aiming for no py2 code in v16.0. Although unfortunately I don't think that we have the manpower to make that a high priority item and ensure that it happens smoothly. So I think that we'll need to consider it carefully and plan our attack.

There may also be good reason to completely rewrite at least some of it. Especially if we move to a (docker style) container appliance model (which we have been considering for years - but never made any moves on). The base (docker style) container that we've been playing with is a super minimalist Debian and weighs in at ~20MB. It doesn't include Python at all (and more than doubles the size if we do)! So there may be a good case to rewrite some applications (e.g. TKLBAM) in something that provides compiled binaries (possibly [Go](https://golang.org/)?).

There is also a case to combine inithooks and confconsole functionality IMO. The core functionality is the same - system configuration. The main difference is when they run. It would also make sense to look around at existing tools and not reinvent the wheel (e.g. leveraging Ansible might make sense?!). Also cloud-init has become something of a defacto standard for pre-configuring servers. Many users in the cloud space expect it (e.g. Azure requires it, Proxmox and AWS can use it, OpenStack uses it by default, etc) so closer integration with that would also make sense.

Bottom line is that it is highly unlikely (let's say it won't happen) that any of the TurnKey specific applications/projects (i.e. code within the ['turnkeylinux' GH org](https://github.com/turnkeylinux)) will exclusively move to py3 within v15.x.

We really need to make a firm decision on whether we stick with the monolithic builds for v16.x, or move to docker style containers (and work out how to integrate that for LXC builds) or perhaps even some sort of hybrid. Then we can have a firm plan of which software we update, which software we rewrite and which software we may even ditch altogether...

For any software that we plan to keep maintaining, then we should at least be aiming to have python3 builds available for v15.x. So then users can assist with testing ASAP.

FWIW, we have already started migrating some of the build tools to py3 (for Buster/v16.x) due to necessity.

> FWIW: I'm thinking there needs to be a transition period when both python2 and python3 are supported.

Agreed. Essentially my plan is to switch any applications themselves (which support it) to py3 by default ASAP (i.e. within v15.x). I'd need to check, but that would be (at least) Ansilbe and web2py. Mayan-EDMS and the Django appliance are already shipping as python3.

That should only leave TurnKey tools as python2. As noted above, we really need to be clear on our longer term plans before it's worth expending too much effort on those. But we need to prioritise that decision making as it is something of a blocker for development (of anything other than Core) for v16.0 really...

> Then in 2020 when python2 is no longer supported, we can finally remove it. Some applications may have to be deprecated if they haven't been upgraded to v3.

Keep in mind that Debian will maintain python2 (in Buster) until 2024, so strictly speaking there is no urgency. If Debian continue with their current release cycle (and no reason to suspect they won't) we can probably expect Bullseye sometime mid 2021 (with no py2 support).

So whilst ideally we should be aiming for no py2 code by v16.0/Buster, we have until v17.0/Bullseye strictly speaking.

**Comment by JedMeister** [comment]
Oops, just saw your new comment. Sorry it's taken me a while to respond... (what I just posted has been in draft pretty much since your last post). I'll read your new post and respond ASAP.

**Comment by JedMeister** [comment]
Regarding your experience with making Ansible scripts for both py2 & py3, perhaps we should just do what we've done with the other appliances and move it to python3. I suspect that there may be the some TurnKey Ansible users who has existing scripts that might like py2 (so it's a very valid aim) but I suspect that many TurnKey users may also be new to Ansible, in which case they should probably be starting off with py3 anyway?!

FWIW the Django appliance is now exclusively py3 (not withstanding the TurnKey py2 tools) since v15.0 IIRC; and we've had no complaints.

**Comment by Dude4Linux** [comment]
From Ansible's Python Compatibility document:
> Ansible has dropped compatibility with Python-2.6 on the controller (The host where /usr/bin/ansible or /usr/bin/ansible-playbook is run). Modules shipped with Ansible can still be used to manage hosts which only have Python-2.6. You just need to have a host with Python-2.7 or Python-3.5 or greater to manage those hosts from.
> 
> One thing that this does affect is the ability to use /usr/bin/ansible-pull to manage a host which has Python-2.6. ansible-pull runs on the host being managed but it is a controller script, not a module so it will need an updated Python. Actively developed Linux distros which ship with Python-2.6 have some means to install newer Python versions (For instance, you can install Python-2.7 via an SCL on RHEL-6) but you may need to also install Python bindings for many common modules to work (For RHEL-6, for instance, selinux bindings and yum would have to be installed for the updated Python install).
> 
> The decision to drop Python-2.6 support on the controller was made because many dependent libraries are becoming unavailable there. In particular, python-cryptography is no longer available for Python-2.6 and the last release of pycrypto (the alternative to python-cryptography) has known security bugs which will never be fixed.

My understanding from this is that we can install and run the Ansible controller (appliance) strictly using Python 3.5 and remove `python-pip` and all the other Python2 modules that were added to support the Ansible app.  The Ansible module, `turnkey_facts.py` and plugin, `lxc-ssh.py` need to be made Python 2/3 compatible so they can run on older Python2 only appliances while being able to run on newer Python3 only appliances.  The `ansible.py` that I referred to is indeed the `inithook` that updates the `ansible user` password during firstboot.  At a minimum, it needs to be converted to Python3 only when the TurnKey py tools are updated.

**Comment by JedMeister** [comment]
> My understanding from this is that we can install and run the Ansible controller (appliance) strictly using Python 3.5 and remove `python-pip` and all the other Python2 modules that were added to support the Ansible app.

Sounds good to me.

> The Ansible module, turnkey_facts.py and plugin, lxc-ssh.py need to be made Python 2/3 compatible so they can run on older Python2 only appliances while being able to run on newer Python3 only appliances.

I'm unfamiliar with Ansilbe, but assuming that they only run on the host, then they should only require py3 support. Although if they also need to be run on the guests, then yes, py2/py3 cross-compatibility would be required to work with older (i.e. < v15.0) TurnKey appliances (and other older distros that do not include Python3 by default).

> The ansible.py that I referred to is indeed the inithook that updates the ansible user password during firstboot. At a minimum, it needs to be converted to Python3 only when the TurnKey py tools are updated.

Agreed.

**Comment by Dude4Linux** [comment]
@JedMeister - Ansible Py3 update:  I've been able to strip out all Python2 modules used by Ansible, replacing them with the equivalent Python3 modules.  I also removed the code which compiled and installed `pycrypto` as it has been deprecated and no longer used by Ansible.

I reverted the change to make `/usr/lib/inithooks/bin/ansible.py` Python 2/3 compatible after I realized it added a dependency on `python-future`.  After thinking about this issue and several other threads, I've come to the conclusion that it makes sense to coordinate the update with the rest of the TurnKey tools & libs, and make that jump to pure Python3 skipping the Python 2/3 compatibility stage.

I've updated the Ansible version to 2.7.7 which is currently the latest stable version.  Ansible modules are written in Python and executed on the target (managed) hosts, therefore they must be Python 2/3 compatible.  Ansible docs have information on how to write them.  The `turnkey_facts.py` implements such a module which gives Ansible the ability to determine if a managed host is running TurnKey GNU/Linux and if so, which type of appliance, version, architecture, etc.  I've re-written the module following Ansible's guidelines and in the process had to change from running `turnkey-version` in a subprocess to just reading the `/etc/turnkey_version` file if it exists.  Makes the module much cleaner.

Ansible plugins are similar to modules in that they are blocks of Python code which add functionality to Ansible.  The main difference is that plugins run only on the Ansible controller (appliance) and therefore can and should be converted to Python3 without backward compatibility.  Since the beginning, our Ansible appliance has included a connection plugin written by Pierre Chifflier which allows Ansible to manage a remote LXC appliance over an `ssh` connection.  It is a blend of the Ansible supplied plugins `ssh.py` and `lxc.py`.  AFAIK, Pierre hasn't updated the plugin since it's first release.

I added a fix to copy `/etc/skel-ansible` to the `/home/ansible` directory.  Apparently my last 'fix' of adding `SKEL=/etc/skel-ansible` to the command line is ignored by `adduser`.  I tried using `useradd` with the -k option, but that led to other problems setting up `ssh-agent`.  Also updated the `usage.rst` file.

TODO:

1. Re-write connection plugin `lxc_ssh.py` for Python3 compatibility and Ansible 2.7 API.
2. Update `init.yml` playbook so hosts are initialized for Python2 or Python3 as appropriate.
3. Test ability to manage LXC containers on both v14.2 and v15.x appliances.

**Comment by JedMeister** [comment]
@Dude4Linux - Great work mate. Sounds awesome.

> I've been able to strip out all Python2 modules used by Ansible, replacing them with the equivalent Python3 modules. I also removed the code which compiled and installed pycrypto as it has been deprecated and no longer used by Ansible.

Brilliant!

> I reverted the change to make /usr/lib/inithooks/bin/ansible.py Python 2/3 compatible after I realized it added a dependency on python-future. After thinking about this issue and several other threads, I've come to the conclusion that it makes sense to coordinate the update with the rest of the TurnKey tools & libs, and make that jump to pure Python3 skipping the Python 2/3 compatibility stage.

Agreed. That makes tons of sense. As I noted, Ideally I'd like to see that happen for v16.0, but don't plan to invest too much energy at this point.

> I've updated the Ansible version to 2.7.7 which is currently the latest stable version. Ansible modules are written in Python and executed on the target (managed) hosts, therefore they must be Python 2/3 compatible. Ansible docs have information on how to write them. 

Awesome!

> The turnkey_facts.py implements such a module which gives Ansible the ability to determine if a managed host is running TurnKey GNU/Linux and if so, which type of appliance, version, architecture, etc. I've re-written the module following Ansible's guidelines and in the process had to change from running turnkey-version in a subprocess to just reading the /etc/turnkey_version file if it exists. Makes the module much cleaner.

Ah ok. Yeah, seems like a very reasonable and rational way to go. IIRC turnkey-version is a pretty basic script which essentially does that anyway, so one less call to subprocess is a good thing! 

> Ansible plugins are similar to modules in that they are blocks of Python code which add functionality to Ansible. The main difference is that plugins run only on the Ansible controller (appliance) and therefore can and should be converted to Python3 without backward compatibility. 

Cool, thanks for the Ansible context of this. I had figured that it was something like that, but appreciate having that clarified.

> Since the beginning, our Ansible appliance has included a connection plugin written by Pierre Chifflier which allows Ansible to manage a remote LXC appliance over an ssh connection. It is a blend of the Ansible supplied plugins ssh.py and lxc.py. AFAIK, Pierre hasn't updated the plugin since it's first release.

If it's useful and we continue to use it, perhaps we could fork it and maintain it ourselves? If it's on GitHub, I'm happy to fork it into the 'turnkeylinux' GitHub organisation. OTOH, if it explicitly leverages the code from 2 existing Ansilbe plugins (i.e. by importing them as modules) then perhaps it doesn't need much maintenance?

> I added a fix to copy /etc/skel-ansible to the /home/ansible directory. Apparently my last 'fix' of adding SKEL=/etc/skel-ansible to the command line is ignored by adduser. I tried using useradd with the -k option, but that led to other problems setting up ssh-agent. Also updated the usage.rst file.

Great work mate. As per always, thanks tons for your work.

Thanks for the additional context and background info re Ansible. It is somewhat as I expected, but great to get a better handle on it from someone who knows how to use it! :smile: It seems like such a cool tool and I'd really love to start using it, but I just never seem to have the time to wrap my head around it... I'm hoping as I learn more about Python and get better at using that, that when I do get next a chance to have a closer look, it will be easier to pick up (than when I've previously glanced at it).

Finally, please correct me if I'm wrong, but we still need to finalise your final dot points before we're ready to rebuild/update. Is that right?!

**Comment by Dude4Linux** [comment]
@JedMeister - Yes, those bullet points need to be completed before it's ready for release.  I should also add taking another look at the usage documents to make sure they cover the new features.  I've bumped the turnkey-version to 15.1.  I'll issue a new pull request when it's ready.  Right now I'm having trouble getting the LXC appliance to run inside a LXD/LXC container.  In theory, it should be working when security.nesting and security.privileged are enabled, but still no joy.  Also I can't make sense of the error messages probably because no one else is trying to do this.

**Comment by JedMeister** [comment]
Sorry, I can't give any advice on running LXC within LXC. As you note, it should be possible, but I've never tried it myself.

Look forward to your PR for Ansible update! :smile:

**Comment by Dude4Linux** [comment]
@JedMeister - I'm currently running LXC and LXD versions 3.0.3 on my laptop with Ubuntu 16.04.6 with an adapted version of the `lxc-turnkey` template that allows me to download and install TurnKey appliances. I tried installing Ubuntu 18.04 into a container and then create a nested container using the latest versions of LXC and LXD included in 18.04.  I got the exact same errors that I received with the nested TurnKey LXC appliance.  This tells me the problem is in my base system.

I think I found reference to a bug that could be causing this behavior.  [Bug#916639](https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1649005.html): LXC AppArmor confinement breaks systemd v240 describes a problem in LXC 3.0.3 that might be related.

I guess my next step is to look for an updated version 3.1 of LXC.
